Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerServlet.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerServlet.java	(revision )
@@ -0,0 +1,37 @@
+package cn.univyz.bfyweb.controller;
+
+import cn.univyz.bfyweb.model.Customer;
+import cn.univyz.bfyweb.service.CustomerService;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * 进人客户列表界面
+ */
+@WebServlet("/Customer")
+public class CustomerServlet extends HttpServlet{
+
+    private CustomerService customerService;
+
+    @Override
+    public void init() throws ServletException{
+        customerService=new CustomerService();
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String keyword="";
+        List<Customer> customersList = customerService.getCustomerList(keyword);
+        req.setAttribute("customerList",customersList);
+        req.getRequestDispatcher("/WEB-INF/view/Customer.jsp").forward(req,resp);
+
+    }
+
+
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/service/CustomerService.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/service/CustomerService.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/service/CustomerService.java	(revision )
@@ -0,0 +1,57 @@
+package cn.univyz.bfyweb.service;
+
+import cn.univyz.bfyweb.helper.DatabaseHelper;
+import cn.univyz.bfyweb.model.Customer;
+import cn.univyz.bfyweb.util.PropsUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * 提供客户数据服务
+ */
+public class CustomerService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PropsUtil.class);
+
+    /**
+     * 获取客户列表
+     */
+    public List<Customer> getCustomerList(String keyword){
+               String sql ="SELECT * FROM Customer";
+               return DatabaseHelper.queryEntityList(Customer.class,sql);
+    }
+
+    /**
+     * 获取客户
+     */
+    public Customer getCustomer(long id){
+        String sql = "SELECT * FROM Customer WHERE id="+id;
+        return DatabaseHelper.queryEntity(Customer.class,sql);
+    }
+    /**
+     * 创建客户
+     */
+    public boolean createCustomer(Map<String,Object> fieldMap){
+        return  DatabaseHelper.insertEntity(Customer.class,fieldMap);
+    }
+
+    /**
+     * 更新客户
+     */
+    public boolean updateCustomer(long id,Map<String,Object> fieldMap){
+
+        return DatabaseHelper.updateEntity(Customer.class, id, fieldMap);
+    }
+    /**
+     * 删除客户
+     */
+    public boolean deleteCustomer(long id){
+
+        return DatabaseHelper.deleteEntity(Customer.class, id);
+    }
+    
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerDeleteServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerDeleteServlet.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerDeleteServlet.java	(revision )
@@ -0,0 +1,18 @@
+package cn.univyz.bfyweb.controller;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+
+/**
+ * 处理删除客户请求
+ */
+public class CustomerDeleteServlet extends HttpServlet{
+    @Override
+    protected void doDelete(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        //TODO
+    }
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/CastUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/CastUtil.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/CastUtil.java	(revision )
@@ -0,0 +1,110 @@
+package cn.univyz.bfyweb.util;
+
+/**
+ * 转型操作工具类
+ */
+public final class CastUtil {
+    /**
+     * 转为String型
+     */
+    public static String castString(Object obj){
+        return CastUtil.castString(obj,"");
+    }
+    /**
+     * 转为String型(提供默认值)
+     */
+    public static String castString(Object obj,String defaultValue){
+        return obj !=null? String.valueOf(obj) : defaultValue;
+    }
+
+    /**
+     * 转为double型
+     */
+        public static double castDouble(Object obj){
+            return CastUtil.castDouble(obj,0);
+        }
+    /**
+     * 转为double型(提供默认值)
+     */
+    public static double castDouble(Object obj,double defaultValue){
+        double doubleValue = defaultValue;
+        if (obj != null) {
+            String strValue=castString(obj);
+            if (StringUtil.isNotEmpty(strValue)){
+                try{
+                    doubleValue=Double.parseDouble(strValue);
+                }catch (NumberFormatException e){
+                    doubleValue=defaultValue;
+                }
+            }
+        }
+        return doubleValue;
+    }
+
+    /**
+     * 转为long型
+     */
+    public static long castLong(Object obj){
+        return CastUtil.castLong(obj,0);
+    }
+    /**
+     * 转为long型(提供默认值)
+     */
+    public static long castLong(Object obj,long defaultValue){
+        long longValue=defaultValue;
+        if (obj !=null){
+            String strValue=castString(obj);
+            if (StringUtil.isNotEmpty(strValue)){
+                try{
+                    longValue=Long.parseLong(strValue);
+                }catch (NumberFormatException e){
+                    longValue=defaultValue;
+                }
+            }
+        }
+        return longValue;
+    }
+
+    /**
+     * int转为型
+     */
+    public static int castInt(Object obj){
+        return CastUtil.castInt(obj,0);
+    }
+
+    /**
+     * 转为int型(提供默认值)
+     */
+    public static int castInt(Object obj,int defaultValue){
+        int intValue=defaultValue;
+        if (obj !=null){
+            String strValue=castString(obj);
+            if (StringUtil.isNotEmpty(strValue)){
+                try{
+                    intValue=Integer.parseInt(strValue);
+                }catch (NumberFormatException e){
+                    intValue=defaultValue;
+                }
+            }
+        }
+        return intValue;
+    }
+    /**
+     * 转为boolean型
+     */
+    public static boolean castBoolean(Object obj){
+        return CastUtil.castBoolean(obj,false);
+    }
+    /**
+     * 转为boolean型(提供默认值)
+     */
+    public static boolean castBoolean(Object obj,boolean defaultValue){
+        boolean booleanValue=defaultValue;
+        if (obj !=null){
+            booleanValue=Boolean.parseBoolean(castString(obj));
+        }
+        return booleanValue;
+    }
+
+
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerCreateServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerCreateServlet.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerCreateServlet.java	(revision )
@@ -0,0 +1,33 @@
+package cn.univyz.bfyweb.controller;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+
+/**
+ * 创建客户
+ */
+@WebServlet("/customer_create")
+public class CustomerCreateServlet extends HttpServlet{
+    /**
+     * 进入创建客户界面
+     */
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        //TODO
+
+    }
+    /**
+     * 处理创建客户请求
+     */
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        //TODO
+
+    }
+}
Index: .IntelliJIdea2017.2/config/options/window.state.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- .IntelliJIdea2017.2/config/options/window.state.xml	(revision )
+++ .IntelliJIdea2017.2/config/options/window.state.xml	(revision )
@@ -0,0 +1,7 @@
+<application>
+  <component name="WindowStateApplicationService">
+    <state x="302" y="203" width="1465" height="691" key="#com.intellij.execution.junit2.states.ComparisonFailureState$DiffDialog" />
+    <state x="302" y="203" width="1465" height="691" key="#com.intellij.execution.junit2.states.ComparisonFailureState$DiffDialog/1920.1080" />
+    <state x="302" y="203" width="1465" height="691" key="#com.intellij.execution.junit2.states.ComparisonFailureState$DiffDialog@1920.1080@144dpi" />
+  </component>
+</application>
\ No newline at end of file
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerEditServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerEditServlet.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerEditServlet.java	(revision )
@@ -0,0 +1,17 @@
+package cn.univyz.bfyweb.controller;
+
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
+ * 处理编辑客户请求
+ */
+public class CustomerEditServlet extends HttpServlet {
+    @Override
+    protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        //TODO
+    }
+}
Index: IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_create.jsp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_create.jsp	(revision )
+++ IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_create.jsp	(revision )
@@ -0,0 +1,19 @@
+<%--
+  Created by IntelliJ IDEA.
+  User: bfy
+  Date: 17-11-2
+  Time: 上午10:22
+  To change this template use File | Settings | File Templates.
+--%>
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>客户管理 - 创建客户</title>
+</head>
+<body>
+<h1>创建客户界面</h1>
+
+<%-- TODO  --%>
+
+</body>
+</html>
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/helper/DatabaseHelper.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/helper/DatabaseHelper.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/helper/DatabaseHelper.java	(revision )
@@ -0,0 +1,233 @@
+package cn.univyz.bfyweb.helper;
+
+import cn.univyz.bfyweb.util.CollectionUtil;
+import cn.univyz.bfyweb.util.PropsUtil;
+import org.apache.commons.dbcp2.BasicDataSource;
+import org.apache.commons.dbutils.QueryRunner;
+import org.apache.commons.dbutils.handlers.BeanHandler;
+import org.apache.commons.dbutils.handlers.BeanListHandler;
+import org.apache.commons.dbutils.handlers.MapListHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.BufferedReader;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.List;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.Map;
+import java.util.Properties;
+
+/**
+ * 数据库操作助手类
+ */
+public final class DatabaseHelper {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DatabaseHelper.class);
+    private static final QueryRunner QUERY_RUNNER;
+    private static final ThreadLocal<Connection> CONNECTION_HOLDER;
+    private static final BasicDataSource DATA_SOURCE;
+    static {
+        CONNECTION_HOLDER = new ThreadLocal<Connection>();
+
+        QUERY_RUNNER = new QueryRunner();
+
+        Properties conf = PropsUtil.loadProps("config.properties");
+        String driver=conf.getProperty("jdbc.driver");
+        String url=conf.getProperty("jdbc.url");
+        String username=conf.getProperty("jdbc.username");
+        String password=conf.getProperty("jdbc.password");
+
+        DATA_SOURCE=new BasicDataSource();
+        DATA_SOURCE.setDriverClassName(driver);
+        DATA_SOURCE.setUrl(url);
+        DATA_SOURCE.setUsername(username);
+        DATA_SOURCE.setPassword(password);
+
+    }
+
+    /**
+     * 获取数据库连接
+     */
+    public  static Connection getConnection() {
+        Connection conn = CONNECTION_HOLDER.get();
+        try{
+            conn= DATA_SOURCE.getConnection();
+        }catch (SQLException e){
+            LOGGER.error("get connection failure",e);
+            throw new RuntimeException(e);
+        }finally {
+            CONNECTION_HOLDER.set(conn);
+        }
+        return conn;
+    }
+
+    /**
+     * 关闭数据库连接
+     */
+    /*
+    public  static void closeConnection(){
+        Connection conn =CONNECTION_HOLDER.get();
+        if (conn != null){
+            try{
+                conn.close();
+            }catch (SQLException e){
+                LOGGER.error("close connection failure",e);
+                throw new RuntimeException(e);
+            }finally {
+                CONNECTION_HOLDER.remove();
+            }
+        }
+    }
+    */
+
+
+    /**
+     * 查询实体列表
+     */
+    public static <T>List<T>  queryEntityList(Class<T> entityClass, String sql, Object... params){
+        List<T> entityList;
+
+        try{
+            Connection conn=getConnection();
+            entityList = QUERY_RUNNER.query(conn,sql,new BeanListHandler<T>(entityClass),params);
+        }catch (SQLException e){
+            LOGGER.error("query entity list failure",e);
+            throw new RuntimeException(e);
+        }
+        return entityList;
+    }
+
+    /**
+     * 查询实体
+     */
+    public static <T>T queryEntity(Class<T> entityClass,String sql,Object... params){
+        T entity;
+
+        try{
+            Connection conn=getConnection();
+            entity = QUERY_RUNNER.query(conn,sql,new BeanHandler<T>(entityClass),params);
+        }catch (SQLException e){
+            LOGGER.error("query entity failure",e);
+            throw new RuntimeException(e);
+        }
+        return entity;
+    }
+
+    /**
+     * 执行查询语句
+     */
+    public static List<Map<String,Object>> executeQuery(String sql,Object... params){
+        List<Map<String,Object>> result;
+        try{
+            Connection conn = getConnection();
+            result=QUERY_RUNNER.query(conn,sql,new MapListHandler(),params);
+        }catch (Exception e){
+            LOGGER.error("execute query failure",e);
+            throw new RuntimeException();
+        }
+        return  result;
+    }
+
+    /**
+     * 执行更新语句（包括update,insert,delete）
+     */
+    public static int executeUpdate(String sql,Object... params){
+        int rows=0;
+        try{
+            Connection conn = getConnection();
+            rows=QUERY_RUNNER.update(conn,sql,params);
+        }catch (SQLException e){
+            LOGGER.error("execute update failure",e);
+            throw new RuntimeException(e);
+        }
+        return rows;
+    }
+
+    /**
+     * 插入实体
+     */
+    public static <T> boolean insertEntity(Class<T> entityClass,Map<String,Object> fieldMap) {
+
+
+        if (CollectionUtil.isEmpty(fieldMap)){
+            LOGGER.error("can not insert entity: fieldMap is empty");
+            return false;
+        }
+
+        String sql="INSERT INTO " + getTableName(entityClass);
+        StringBuilder columns =new StringBuilder("(");
+        StringBuilder values=new StringBuilder("(");
+        for (String fieldName : fieldMap.keySet()){
+            columns.append(fieldName).append(", ");
+            values.append("?, ");
+        }
+        columns.replace(columns.lastIndexOf(", "), columns.length(),")");
+        values.replace(values.lastIndexOf(", "), values.length(), ")");
+        sql += columns + " VALUES " + values;
+        Object[] params =fieldMap.values().toArray();
+        return executeUpdate(sql,params) == 1;
+
+    }
+
+    /**
+     * 更新实体
+     */
+    public static <T> boolean updateEntity(Class<T> entityClass, long id,Map<String,Object> fieldMap){
+        if (CollectionUtil.isEmpty(fieldMap)){
+            LOGGER.error("can not update entity: fieldMap is empty");
+            return false;
+        }
+
+        String sql ="UPDATE " + getTableName(entityClass) + " SET ";
+        StringBuilder columns = new StringBuilder();
+        for (String fieldName : fieldMap.keySet()) {
+            columns.append(fieldName).append("=?, ");
+        }
+        sql += columns.substring(0,columns.lastIndexOf(", ")) + " WHERE id=?";
+        List<Object> paramList = new ArrayList<Object>();
+        paramList.addAll(fieldMap.values());
+        paramList.add(id);
+        Object[] params = paramList.toArray();
+
+
+        return executeUpdate(sql,params) == 1;
+    }
+
+    /**
+     * 删除实体
+     */
+    public static <T> boolean deleteEntity(Class<T> entityClass, long id){
+        String sql="DELETE FROM " + getTableName(entityClass) + " WHERE id=?";
+
+        return executeUpdate(sql,id) ==1;
+    }
+
+    /**
+     * 执行SQL文件
+     * @param filePath
+     */
+    public static void executeSqlFile(String filePath){
+
+        InputStream is=Thread.currentThread().getContextClassLoader().getResourceAsStream(filePath);
+        BufferedReader reader = new BufferedReader(new InputStreamReader(is));
+        String sql;
+
+        try {
+            while((sql=reader.readLine()) !=null){
+                executeUpdate(sql);
+            }
+        }catch (Exception e){
+            LOGGER.error("execute sql file failure", e);
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    public static String getTableName(Class<?> entityClass){
+        return entityClass.getSimpleName().toLowerCase();
+    }
+
+
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerController.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerController.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerController.java	(revision )
@@ -0,0 +1,103 @@
+package cn.univyz.bfyweb.controller;
+/*
+import org.smart4j.chapter3.model.Customer;
+import org.smart4j.chapter3.service.CustomerService;
+import org.smart4j.framework.annotation.Action;
+import org.smart4j.framework.annotation.Controller;
+import org.smart4j.framework.annotation.Inject;
+import org.smart4j.framework.bean.Data;
+import org.smart4j.framework.bean.Param;
+import org.smart4j.framework.bean.View;
+*/
+
+/**
+ * 处理客户管理相关请求
+ */
+//@Controller
+public class CustomerController {
+
+    /*
+    @Inject
+    private CustomerService customerService;
+
+    /**
+     * 进入客户列表界面
+     */
+    /*
+    @Action("get:/Customer")
+    public View index(Param param){
+        List<Customer> customerList = customerService.getCustomerList();
+        return new View("Customer.jsp").addModel("customerList",customerList);
+
+    }
+
+    /**
+     * 显示客户基本信息
+     */
+    /*
+    @Action("get:/customer_show")
+    public View show(Param param){
+        long id=param.getLong("id");
+        Customer Customer = customerService.getCustomer(id);
+        return new View("customer_show.jsp").addModel("Customer",Customer);
+
+    }
+
+    /**
+     * 进入创建客户界面
+     */
+    /*
+    @Action("get:/customer_create")
+    public View create(Param param){
+        return new View("customer_create.jsp");
+    }
+
+    /**
+     * 处理创建客户请求
+     */
+    /*
+    @Action("post:/customer_create")
+    public Data createSubmit(Param param){
+        Map<String,Object> fieldMap=param.getMap();
+        boolean result = customerService.createCustomer(fieldMap);
+        return new Data(result);
+    }
+
+    /**
+     * 进入编辑客户界面
+     */
+    /*
+    @Action("get:/customer_edit")
+    public View edit(Param param){
+        long id = param.getLong("id");
+        Customer Customer=customerService.getCustomer(id);
+        return new View("customer_edit.jsp").addModel("Customer",Customer);
+    }
+
+    /**
+     *  处理编辑客户请求
+     */
+    /*
+    @Action("put:/customer_edit")
+    public View editSubnit(Param param){
+        long id = param.getLong("id");
+        Map<String,Object> fieldMap = param.getMap();
+        boolean result=customerService.updateCustomer(id, fieldMap);
+        return new Data(result);
+    }
+
+    /**
+     * 处理删除客户请求
+     */
+    /*
+    @Action("delete:/customer_edit")
+    public Data delete(Param param){
+        long id = param.getLong("id");
+        boolean result = customerService.deleteCustomer(id);
+        return new Data(result);
+    }
+
+    */
+
+
+}
Index: IdeaProjects/bfyweb/src/test/java/cn/univyz/bfyweb/test/CustomerServiceTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/test/java/cn/univyz/bfyweb/test/CustomerServiceTest.java	(revision )
+++ IdeaProjects/bfyweb/src/test/java/cn/univyz/bfyweb/test/CustomerServiceTest.java	(revision )
@@ -0,0 +1,69 @@
+package cn.univyz.bfyweb.test;
+
+import cn.univyz.bfyweb.helper.DatabaseHelper;
+import cn.univyz.bfyweb.model.Customer;
+import cn.univyz.bfyweb.service.CustomerService;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * 单元测试
+ */
+public class CustomerServiceTest {
+    private final CustomerService customerService;
+
+    public CustomerServiceTest() {
+        this.customerService =new CustomerService();
+
+    }
+
+    @Before
+    public void init() throws Exception {
+        //TODO 初始化数据库
+        DatabaseHelper.executeSqlFile("sql/customer_init.sql");
+    }
+    @Test
+    public void getCustomerListText() throws Exception{
+        String keyword=null;
+        List<Customer> customerList = customerService.getCustomerList(keyword);
+        Assert.assertEquals(2,customerList.size());
+    }
+
+    @Test
+    public  void getCustomerTest() throws Exception{
+        long id=1;
+        Customer customer = customerService.getCustomer(id);
+        Assert.assertNotNull(customer);
+    }
+
+    @Test
+    public void createCustomerTest() throws Exception{
+        Map<String, Object> fieldMap = new HashMap<String, Object>();
+        fieldMap.put("name","customer100");
+        fieldMap.put("contact","bfy32");
+        fieldMap.put("telephone","13512345678");
+        boolean result = customerService.createCustomer(fieldMap);
+        Assert.assertTrue(result);
+    }
+
+    @Test
+    public void updateCustomerTest() throws Exception{
+        long id=1;
+        Map<String, Object> fieldMap = new HashMap<String,Object>();
+        fieldMap.put("contact","Eric");
+        boolean result = customerService.updateCustomer(id,fieldMap);
+        Assert.assertTrue(result);
+    }
+
+    @Test
+    public void deleteCustomerTest() throws Exception{
+        long id=1;
+        boolean result = customerService.deleteCustomer(id);
+        Assert.assertTrue(result);
+    }
+}
Index: IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_edit.jsp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_edit.jsp	(revision )
+++ IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_edit.jsp	(revision )
@@ -0,0 +1,17 @@
+<%--
+  Created by IntelliJ IDEA.
+  User: bfy
+  Date: 17-11-3
+  Time: 下午3:07
+  To change this template use File | Settings | File Templates.
+--%>
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>客服管理 - 编辑客户</title>
+</head>
+<body>
+<h1>编辑客户界面</h1>
+<%-- TODO --%>
+</body>
+</html>
Index: IdeaProjects/bfyweb/src/test/resources/config.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/test/resources/config.properties	(revision )
+++ IdeaProjects/bfyweb/src/test/resources/config.properties	(revision )
@@ -0,0 +1,4 @@
+jdbc.driver=com.mysql.jdbc.Driver
+jdbc.url=jdbc:mysql://localhost:3306/demo_test
+jdbc.username=root
+jdbc.password=lu123456
Index: IdeaProjects/univyz-framework/pom.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/univyz-framework/pom.xml	(revision )
+++ IdeaProjects/univyz-framework/pom.xml	(revision )
@@ -0,0 +1,86 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>cn.univyz</groupId>
+    <artifactId>univyz-framework</artifactId>
+    <version>1.0.0</version>
+
+    <!--添加Servlet,JSP,JSTL等依赖 -->
+    <dependencies>
+
+        <!--Servlet-->
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>3.1.0</version>
+            <scope>provided</scope>
+        </dependency>
+        <!--JSP-->
+        <dependency>
+            <groupId>javax.servlet.jsp</groupId>
+            <artifactId>jsp-api</artifactId>
+            <version>2.2</version>
+            <scope>provided</scope>
+        </dependency>
+        <!--JSTL-->
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>jstl</artifactId>
+            <version>1.2</version>
+            <scope>runtime</scope>
+        </dependency>
+        <!--打印-->
+        <!-- SLF4J -->
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-log4j12</artifactId>
+            <version>1.7.7</version>
+        </dependency>
+        <!-- MySQL -->
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>5.1.43</version>
+            <scope>runtime</scope>
+        </dependency>
+        <!-- Jackson -->
+        <dependency>
+            <groupId>com.fasterxml.jackson.core</groupId>
+            <artifactId>jackson-databind</artifactId>
+            <version>2.4.4</version>
+        </dependency>
+        <!-- 提供常用的工具 -->
+        <!-- Apache Commons Lang -->
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-lang3</artifactId>
+            <version>3.3.2</version>
+        </dependency>
+        <!-- Apache Commons Collections -->
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-collections4</artifactId>
+            <version>4.0</version>
+        </dependency>
+
+        <!--Apache Commons DbUtils -->
+        <dependency>
+            <groupId>commons-dbutils</groupId>
+            <artifactId>commons-dbutils</artifactId>
+            <version>1.6</version>
+        </dependency>
+
+        <!-- 数据库连接池 -->
+        <!-- Apache DBCP -->
+        <dependency>
+            <groupId>org.apache.commons</groupId>
+            <artifactId>commons-dbcp2</artifactId>
+            <version>2.0.1</version>
+        </dependency>
+
+
+    </dependencies>
+</project>
\ No newline at end of file
Index: IdeaProjects/bfyweb/src/main/resources/log4j.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/resources/log4j.properties	(revision )
+++ IdeaProjects/bfyweb/src/main/resources/log4j.properties	(revision )
@@ -0,0 +1,18 @@
+log4j.rootLogger=ERROR,console,file
+
+log4j.appender.console=org.apache.log4j.ConsoleAppender
+log4j.appender.console.layout=org.apache.log4j.PatternLayout
+log4j.appender.console.layout.ConversionPattern=%m%n
+
+
+log4j.appender.file=org.apache.log4j.DailyRollingFileAppender
+log4j.appender.file.File=${user.home}/logs/book.log
+log4j.appender.file.DatePattern='_'yyyyMMdd
+log4j.appender.file.layout=org.apache.log4j.PatternLayout
+log4j.appender.file.layout.ConversionPattern=%d{HH:mm:ss,SSS} %p %c (%L) %m%n
+
+
+log4j.logger.cn.univyz=DEBUG
+
+#将日志级别设置为ERROR,并且提供了两种日志appender,分别console与file.
+#指定只有cn.univyz包下的类才能输出DEBUG级别的日志
Index: IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_show.jsp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_show.jsp	(revision )
+++ IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer_show.jsp	(revision )
@@ -0,0 +1,17 @@
+<%--
+  Created by IntelliJ IDEA.
+  User: bfy
+  Date: 17-11-3
+  Time: 下午3:06
+  To change this template use File | Settings | File Templates.
+--%>
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<html>
+<head>
+    <title>客户管理 - 查看客户</title>
+</head>
+<body>
+<h1>查看客户界面</h1>
+<%-- TODO --%>
+</body>
+</html>
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/StringUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/StringUtil.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/StringUtil.java	(revision )
@@ -0,0 +1,27 @@
+package cn.univyz.bfyweb.util;
+
+
+
+import org.apache.commons.lang3.StringUtils;
+/**
+ * 字符串工具类
+ */
+public final class StringUtil {
+
+    /**
+     * 判断字符串是否为空
+     */
+    public static boolean isEmpty(String str){
+        if (str != null){
+            str = str.trim();
+        }
+        return StringUtils.isEmpty(str);
+    }
+    /**
+     * 判断字符串是否非空
+     */
+    public static  boolean isNotEmpty(String str){
+        return  !isEmpty(str);
+    }
+
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/CollectionUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/CollectionUtil.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/CollectionUtil.java	(revision )
@@ -0,0 +1,35 @@
+package cn.univyz.bfyweb.util;
+
+import org.apache.commons.collections4.MapUtils;
+
+import java.util.Collection;
+import java.util.Map;
+
+public final class CollectionUtil {
+    /**
+     * 判断Collection 是否为空
+     */
+    public static boolean isEmpty(Collection<?> collection){
+        return CollectionUtil.isEmpty(collection);
+    }
+
+    /**
+     * 判断Collection 是否非空
+     */
+
+    public static boolean isNotEmpty(Collection<?> collection){
+        return !isEmpty(collection);
+    }
+    /**
+     * 判断Map 是否为空
+     */
+    public static boolean isEmpty(Map<?,?> map){
+        return MapUtils.isEmpty(map);
+    }
+    /**
+     * 判断Map 是否非空
+     */
+    public static boolean isNotEmpty(Map<?,?> map){
+        return !isEmpty(map);
+    }
+}
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerShowServlet.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerShowServlet.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/controller/CustomerShowServlet.java	(revision )
@@ -0,0 +1,21 @@
+package cn.univyz.bfyweb.controller;
+
+import javax.servlet.ServletException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+
+/**
+ * 显示客户基本信息
+ */
+@WebServlet("/customer_show")
+public class CustomerShowServlet extends HttpServlet{
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        //TODO
+    }
+}
Index: IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer.jsp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer.jsp	(revision )
+++ IdeaProjects/bfyweb/src/main/webapp/WEB-INF/view/customer.jsp	(revision )
@@ -0,0 +1,42 @@
+<%--
+  Created by IntelliJ IDEA.
+  User: bfy
+  Date: 17-11-3
+  Time: 下午12:09
+  To change this template use File | Settings | File Templates.
+--%>
+<%@ page contentType="text/html;charset=UTF-8" language="java" %>
+<%@ taglib prefix="c" uri="http://java.sun.com/jsp/jstl/core" %>
+<c:set var="BASE" value="${pageContext.request.contextPath}"/>
+
+<html>
+<head>
+    <title>客户管理界面</title>
+</head>
+<body>
+
+<h1>客户理表</h1>
+
+<table>
+    <th>客户名称</th>
+    <th>联系人</th>
+    <th>电话号码</th>
+    <th>邮箱地址</th>
+    <th>操作</th>
+    <c:forEach var="customer" items="${customerList}">
+        <tr>
+            <td>${customer.name} </td>
+            <td>${customer.contact} </td>
+            <td>${customer.telephone} </td>
+            <td>${customer.email} </td>
+            <td>
+                <a href="${BASE}/customer_edit?id=${customer.id}">编辑</a>
+                <a href="${BASE}/customer_delete?id=${customer.id}">删除</a>
+            </td>
+        </tr>
+
+    </c:forEach>
+</table>
+
+</body>
+</html>
Index: IdeaProjects/bfyweb/src/test/resources/sql/customer_init.sql
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/test/resources/sql/customer_init.sql	(revision )
+++ IdeaProjects/bfyweb/src/test/resources/sql/customer_init.sql	(revision )
@@ -0,0 +1,4 @@
+TRUNCATE customer;
+INSERT INTO customer ( name, contact, telephone, email, remark) VALUES ('rrrrr','rrrrrr','13500001111','luj@gmail.com',NULL);
+INSERT INTO customer ( name, contact, telephone, email, remark) VALUES ('tttttt','ttttt','13500001111','luj@gmail.com',null);
+commit;
\ No newline at end of file
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/PropsUtil.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/PropsUtil.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/util/PropsUtil.java	(revision )
@@ -0,0 +1,102 @@
+package cn.univyz.bfyweb.util;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+/**
+ * 属性文件工具类
+ */
+public final class PropsUtil {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PropsUtil.class);
+
+    /**
+     * 加载属性文件
+     */
+    public  static Properties loadProps(String fileName){
+        Properties props = null;
+        InputStream is = null;
+        try{
+            is = Thread.currentThread().getContextClassLoader().getResourceAsStream(fileName);
+            if (is ==null){
+                throw new FileNotFoundException(fileName+"file is not found");
+            }
+            props = new Properties();
+            props.load(is);
+        }catch (IOException e){
+            LOGGER.error("load properties file failure",e);
+        }finally {
+            if (is !=null){
+                try{
+                    is.close();
+                }catch (IOException e){
+                    LOGGER.error("close input stream failure", e);
+                }
+            }
+        }
+        return props;
+    }
+
+    /**
+     * 获取字符型属性（默认值为空客串）
+     */
+    public static String getString(Properties props,String key){
+        return getString(props,key,"");
+    }
+
+
+    /**
+     * 获取字符型属性（可指定默认值）
+     */
+    public static String getString(Properties props,String key,String defaultValue){
+        String value = defaultValue;
+        if (props.containsKey(key)){
+            value = props.getProperty(key);
+        }
+        return value;
+    }
+
+    /**
+     * 获取数值型属性（默认值为0）
+     */
+    public static int getInt(Properties props,String key){
+        return getInt(props,key,0);
+    }
+
+    /**
+     * 获取数值型属性（可指定默认值）
+     */
+    public static int getInt(Properties props,String key,int defaultValue){
+        int value = defaultValue;
+        if (props.containsKey(key)){
+            value = CastUtil.castInt(props.getProperty(key));
+        }
+        return value;
+    }
+
+    /**
+     * 获取布尔型属性（默认值为false）
+     */
+    public static boolean getBoolean(Properties props,String key){
+        return getBoolean(props,key,false);
+    }
+
+    /**
+     * 获取布尔型属性（可指定默认值）
+     */
+
+    public static boolean getBoolean(Properties props, String key, Boolean defaultValue){
+        boolean value = defaultValue;
+        if (props.containsKey(key)){
+            value = CastUtil.castBoolean(props.getProperty(key));
+        }
+        return value;
+    }
+
+}
Index: IdeaProjects/bfyweb/src/main/resources/config.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/resources/config.properties	(revision )
+++ IdeaProjects/bfyweb/src/main/resources/config.properties	(revision )
@@ -0,0 +1,4 @@
+jdbc.driver=com.mysql.jdbc.Driver
+jdbc.url=jdbc:mysql://localhost:3306/demo
+jdbc.username=root
+jdbc.password=lu123456
Index: IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/model/Customer.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/model/Customer.java	(revision )
+++ IdeaProjects/bfyweb/src/main/java/cn/univyz/bfyweb/model/Customer.java	(revision )
@@ -0,0 +1,83 @@
+package cn.univyz.bfyweb.model;
+
+/**
+ * 客户
+ */
+public class Customer {
+    /**
+     * ID
+     */
+    private long id;
+
+    /**
+     * 客户名称
+     */
+    private String name;
+
+    /**
+     * 联系人
+     */
+    private String contact;
+
+    /**
+     * 电话号码
+     */
+    private String telephone;
+
+    /**
+     * 邮箱地址
+     */
+    private String email;
+    /**
+     * 备注
+     */
+    private String remark;
+
+    public long getId() {
+        return id;
+    }
+
+    public void setId(long id) {
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
+    }
+
+    public String getContact() {
+        return contact;
+    }
+
+    public void setContact(String contact) {
+        this.contact = contact;
+    }
+
+    public String getTelephone() {
+        return telephone;
+    }
+
+    public void setTelephone(String telephone) {
+        this.telephone = telephone;
+    }
+
+    public String getEmail() {
+        return email;
+    }
+
+    public void setEmail(String email) {
+        this.email = email;
+    }
+
+    public String getRemark() {
+        return remark;
+    }
+
+    public void setRemark(String remark) {
+        this.remark = remark;
+    }
+}
Index: IdeaProjects/bfyweb/src/main/webapp/WEB-INF/web.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- IdeaProjects/bfyweb/src/main/webapp/WEB-INF/web.xml	(revision )
+++ IdeaProjects/bfyweb/src/main/webapp/WEB-INF/web.xml	(revision )
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<web-app xmlns="http://java.sun.com/xml/ns/javaee"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://java.sun.com/xml/ns/javaee
+         http://java.sun.com/xml/ns/javaee/web-app_2_4.xsd"
+         version="3.0">
+</web-app>
+
